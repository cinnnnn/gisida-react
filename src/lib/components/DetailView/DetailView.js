import React, { Component } from 'react';
import { Actions, parseMustache } from 'gisida';
import { connect } from 'react-redux';
import Parser from 'html-react-parser';
import './DetailView.scss';

const mapStateToProps = (state, ownProps) => {
  const { detailView } = state.MAP;
  const layerObj = (detailView && detailView.layerId)
    ? state.MAP.layers[detailView.layerId] : null;
  return {
    MAP: state.MAP,
    layerObj,
    detailView: detailView && detailView.model,
    properties: detailView && detailView.properties,
    spec: detailView && detailView.spec, 
  };
}

class DetailView extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  componentWillReceiveProps(nextProps) {
    const { layerObj, properties, spec, detailView } = nextProps;
    if (!layerObj || !spec || !properties || !detailView) {
      this.setState({ UID: null });
    } else {
      const { UID, title, subTitle, basicInfo, parsedBasicInfo } = detailView;
      this.setState({
        UID,
        title,
        subTitle,
        parsedBasicInfo,
        spec,
        properties,
        basicInfo,
        layerObj,
      });
    }
  }

  onCloseClick(e) {
    e.preventDefault();
    this.props.dispatch(Actions.detailView(null));
  }

  render() {
    const { UID, spec, title, subTitle, parsedBasicInfo } = this.state;
    if (this.props.MAP.showFilterPanel || !UID || !spec) return null;

    const detailList = [];
    if (spec['basic-info']) {
      let detail;
      for (let i = 0; i < parsedBasicInfo.length; i += 1) {
        detail = parsedBasicInfo[i];
        detailList.push((
          <li key={i}>
            <i data-balloon={detail.alt} data-balloon-pos="up">
              <span
                className={`glyphicon glyphicon-${detail.icon}`}
                style={detail.iconColor ? { color: detail.iconColor } : {}}
              />
            </i>
            {typeof detail.value !== 'string' && detail.value.parser ?
              Parser(detail.value)
            : (
              <span>{detail.value}</span>
            )}
          </li>
        ));
      }
    }

    return (
      <div className="detail-view-wrapper">
        <a
          className="close-btn"
          title="Close profile view"
          onClick={e => this.onCloseClick(e)}
          href="/detail-view"
        >
          <span className="glyphicon glyphicon-remove" />
        </a>

        <div className="detail-basic-details">
          <div className="detail-header">
            <h4>{title}</h4>
            {!!subTitle ? (<h6>{subTitle}</h6>) : ''}
          </div>
          <div className="detail-list">
            <ul>{detailList}</ul>
          </div>
        </div>
      </div>
    );
  }
}

export default connect(mapStateToProps)(DetailView);
