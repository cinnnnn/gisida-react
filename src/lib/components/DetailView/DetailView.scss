@import url(//cdnjs.cloudflare.com/ajax/libs/balloon-css/0.5.0/balloon.min.css);

.detail-view-wrapper {
  background: #f7f6ef;
  box-shadow: 0 1px 7px rgba(0, 0, 0, 0.4);
  font-size: 0.9em;
  font-family: proxima-nova, sans-serif;
  line-height: 20px;
  margin: 0.2em;
  position: absolute;
  z-index: 1001;
  right: 0;
  top: 0;
  color: #4c4c4c;
  letter-spacing: 0.01em;
  bottom: 0;
  width: 345px;
  position: fixed;
  height: 100%;
  display: flex;
  flex-flow: column nowrap;

  .close-btn {
    position: absolute;
    right: 15px;
    top: 7px;
    z-index: 1000;
    color: #555;
    display: block;
  }

  .detail-basic-details {
    .detail-header {
      min-height: 90px;
      margin: 0.4em 0;
      border-bottom: 1px solid #e8e8e8;
      line-height: 1.3;
      padding: 0 10px;

      h4 {
        position: relative;
        color: rgb(0, 153, 255);
        margin: 0.6em 0 0;
        line-height: 1.3em;
        padding: 0;
        font-weight: 600;
      }

      h6 {
        margin-top: 0;
      }

      img {
        width: 70px;
        height: 45px;
        padding: 0 1px;
        margin-bottom: 10px;
      }
    }

    .detail-list {
      min-height: 90px;
      margin: 0.1em 0;
      border-bottom: 1px solid #e8e8e8;
      line-height: 1.3;
      padding: 0 10px;

      ul,
      li {
        padding: 1px;
        margin: 0;
        list-style: none;
      }

      li {
        margin: 5px 0;
        font-weight: normal;
        font-size: 12px;
        color: #555;

        i {
          float: left;
          margin-right: 10px;
          color: #09f;
        }
      }
    }
    .indicators-section-header {
      padding: 10px 12px;
  
      a {
        width: 100%;
        padding: 0 !important;
  
        .pie-chart-legend {
          position: relative;
          float: right;
          width: 44%;
          top: 24px;
  
          ul {
            padding: 0 !important;
  
            li {
              text-indent: -30px;
  
              span {
                left: 0 !important;
                top: 6px;
              }
            }
          }
  
          td {
            font-weight: 500;
            padding-top: 10px;
  
            &:first-child {
              text-align: center;
            }
          }
  
          img {
            width: 20px;
            height: 20px;
  
            &.boy,
            &.boys {
              content: url('/assets/img/icon-boy.svg');
            }
  
            &.girl,
            &.girls {
              content: url('/assets/img/icon-girl.svg');
            }
  
            &.pregnant,
            &.pregnant {
              content: url('/assets/img/icon-pregnant.svg');
            }
          }
        }
      }
  
      .rpNav {
        cursor: pointer;
        top: 1px;
        position: relative;
        font-size: 0.9em;
  
        &.glyphicon-triangle-left {
          padding-left: 6px;
        }
  
        &.disabled {
          cursor: auto;
          color: #aaa;
        }
      }
  
      .rpSelect {
        border: none;
        background-color: transparent;
  
        &:focus {
          outline: none;
        }
      }
    }

    .profile-indicators {
      flex-shrink: 1;
      margin: 0;
      border-bottom: 1px solid #e8e8e8;
      line-height: 1.3;
      padding: 0;
      background: #f5f3e9;
      overflow: auto;
      flex-basis: 70%;
      height: 70%;
  
      ul {
        list-style-type: none;
        padding: 1px;
        margin-top: 0;
  
        li {
          list-style-type: none;
  
          &.indicators-section-header {
            div.progress-container {
              width: 100%;
              height: 80px;
              position: relative;
              margin-top: -20px;
  
              div.progress-bar-labels {
                width: 80%;
                float: right;
                margin-right: 30px;
              }
  
              div.progress-value {
                position: absolute;
                right: 0;
                top: 54px;
              }
  
              div.progress-bar-icon {
                width: 10%;
                position: relative;
                top: 50%;
  
                img {
                  width: 20px;
                  height: 30px;
  
                  &.pregnant {
                    content: url('/assets/img/icon-pregnant.svg');
                  }
  
                  &.boy,
                  &.boys {
                    content: url('/assets/img/icon-boy.svg');
                  }
  
                  &.girl,
                  &.girls {
                    content: url('/assets/img/icon-girl.svg');
                  }
                }
              }
            }
  
            div.service-parameters {
              display: block;
  
              img {
                float: none;
                width: 30%;
                left: 0;
                padding-right: 40px;
                padding-left: 40px;
                display: inline-block;
              }
            }
          }
  
          .indicator-label {
            font-weight: 500;
  
            span.indicator-totals {
              left: -10px !important;
              bottom: 16px;
              position: relative;
            }
  
            span.indicator-value {
              float: right;
              left: -10px;
              width: 40px;
              text-align: left;
              font-weight: 500 !important;
            }
  
            span.caret-down,
            span.caret-right {
              right: 30px;
              top: 10px;
              position: absolute;
  
              &.total {
                top: 35px;
              }
            }
  
            span.caret-right {
              transform: rotate(-90deg);
            }
  
            span.fa-square {
              left: 0;
              margin-right: 10px;
              vertical-align: middle;
            }
          }
  
          ul.indicator-bar {
            padding: 0;
          }
  
          .indicator-item {
            display: block;
            padding-left: 30px;
            position: relative;
  
            span.caret-down,
            span.caret-right {
              bottom: 0;
            }
  
            &::before {
              content: '';
              height: 12px;
              width: 12px;
              position: absolute;
              left: 12px;
            }
  
            &.Recovered::before {
              background-color: #2ecc40;
            }
  
            &.Non-recovered::before {
              background-color: #ffdc00;
            }
  
            &.Defaulted::before {
              background-color: orange;
            }
  
            &.Died::before {
              background-color: red;
            }
          }
  
          .service-name {
            font-weight: 600 !important;
          }
  
          a {
            color: #555;
            width: 100%;
            display: block;
            font-weight: 600;
            padding: 10px 12px;
            border-bottom: 1px solid #eee;
            position: relative;
  
            span {
              &:first-child {
                left: 70px;
                position: relative;
              }
  
              &:last-child {
                float: right;
                position: relative;
                margin-top: 5px;
              }
            }
  
            &:hover,
            &:active,
            &:focus {
              text-decoration: none;
            }
          }
  
          ul {
            padding: 20px 0 20px 20px;
          }
        }
      }
    }
    .progress {
      background-color: #f9f9f9;
      border-radius: 0;
      color: #fff;
      font-size: 11px;
      font-size: 0.6875rem;
      height: 20px;
      line-height: 1.182;
      margin: 6px 0;
      position: relative;
      text-align: center;
      width: 80%;
      float: right;
      margin-right: 30px;
      overflow: inherit !important;
  
      &.progress-stacked {
        width: 100% !important;
        float: none !important;
        display: -webkit-box;
      }
  
      & > span {
        position: relative;
        float: left !important;
        height: 100%;
        margin: auto !important;
        left: 0 !important;
        pointer-events: all;
      }
    }
  }

  .detail-extension-wrapper {
    overflow-y: auto;
  }
}

[data-balloon]::after {
  white-space: inherit !important;
  padding: 0.5em !important;
  width: 100px;
  text-align: center;
}

#facilityImg {
  border-radius: 5px;
  cursor: -moz-zoom-in;
  cursor: -webkit-zoom-in;
  cursor: zoom-in;
  transition: 0.3s;
}

#facilityImg:hover { opacity: 0.7; }

/* The Modal (background) */
.modal {
  display: none; /* Hidden by default */
  position: fixed; /* Stay in place */
  z-index: 1; /* Sit on top */
  padding-top: 100px; /* Location of the box */
  left: 0;
  top: 0;
  width: 100%; /* Full width */
  height: 100%; /* Full height */
  overflow: auto; /* Enable scroll if needed */
  background-color: rgb(0, 0, 0); /* Fallback color */
  background-color: rgba(37, 37, 37, 0.9); /* Black w/ opacity */
}

/* Modal Content (image) */
.modal-content {
  margin: auto;
  display: block;
  width: 80%;
  max-width: 500px;
  padding: 0 !important;
}

/* Caption of Modal Image */
#caption {
  margin: auto;
  display: block;
  width: 80%;
  max-width: 700px;
  text-align: center;
  color: #fff;
  padding: 10px 0;
}

/* Add Animation */
.modal-content,
#caption {
  -webkit-animation-name: zoom;
  -webkit-animation-duration: 0.6s;
  animation-name: zoom;
  animation-duration: 0.6s;
}

@-webkit-keyframes zoom {
  from { -webkit-transform: scale(0); }
  to { -webkit-transform: scale(1); }
}

@keyframes zoom {
  from { transform: scale(0); }
  to { transform: scale(1); }
}

/* The Close Button */
.close {
  position: absolute;
  top: 15px;
  right: 35px;
  color: #fff !important;
  font-size: 40px;
  font-weight: bold;
  transition: 0.3s !important;
  opacity: 0.8 !important;
}

.close:hover,
.close:focus {
  color: #bbb;
  text-decoration: none;
  cursor: pointer;
}

/* 100% Image Width on Smaller Screens */
@media only screen and (max-width: 700px) {
  .modal-content {
    width: 100%;
  }
}
