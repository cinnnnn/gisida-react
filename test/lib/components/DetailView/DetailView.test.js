import React from 'react';
import DetailView from '../../../../src/lib/components/DetailView/DetailView';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';

const initialState = {
}

const mockStore = configureStore()
const store = mockStore(initialState);


const componentWrapper = mount(
	<DetailView 
	store={store}
	  MAP={{timeseries: {visibility:true}}}
	/>

	);

describe('DetailView', () => {
	it('DetailView component renders correctly', () => {
		const json = toJson(componentWrapper)
		expect(json).toMatchSnapshot();
	})

	it('should set props', () => {
		const propsUpdateMock = jest.spyOn(componentWrapper.instance(), 'componentWillReceiveProps');
		componentWrapper.setProps({bottom: 62});
		expect(componentWrapper.props().bottom).toEqual(62);
		expect(propsUpdateMock).toHaveBeenCalledTimes(1);
	})
});