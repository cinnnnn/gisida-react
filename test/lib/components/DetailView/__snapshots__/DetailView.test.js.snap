// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailView DetailView component renders correctly 1`] = `
<Connect(DetailView)
  MAP={
    Object {
      "timeseries": Object {
        "visibility": true,
      },
    }
  }
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <DetailView
    MAP={
      Object {
        "filter": Object {},
        "layers": Object {},
        "timeseries": Object {},
      }
    }
    dispatch={[Function]}
    layerObj={null}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
  />
</Connect(DetailView)>
`;
